src/dns/query.cpp

```diff
@@ -3,46 +3,306 @@
+            size = 4;
+            logger->info("dns - IPv4");
+        } else {
+            family = PF_INET6;
+            size = 16;
+            logger->info("dns - IPv6");
+        }
+
+        //
+        // Note: make sure in advance `i * size` won't overflow,
+        // this is here only for robustness.
+        //
+        if (count >= 0 && count <= INT_MAX / size + 1) {
+
+            for (auto i = 0; i < count; ++i) {
+                // Note: address already in network byte order
+                if (inet_ntop(family, (char *)addresses + i * size,
+                              string, sizeof(string)) == NULL) {
+                    logger->warn("dns - unexpected inet_ntop failure");
+                    code = DNS_ERR_UNKNOWN;
+                    break;
+                }
+                Answer answer;
+                answer.code = code;
+                answer.ttl = ttl;
+                if (family == PF_INET) {
+                    answer.ipv4 = string;
+                    answer.type = QueryTypeId::A;
+                } else if (family == PF_INET6) {
+                    answer.ipv6 = string;
+                    answer.type = QueryTypeId::AAAA;
+                }
+                logger->info("dns - adding '%s'", string);
+                answers.push_back(answer);
+            }
+
+        } else {
+            logger->warn("dns - too many addresses");
+            code = DNS_ERR_UNKNOWN;
+        }
+
+    } else {
+        logger->warn("dns - invalid response type");
+        code = DNS_ERR_UNKNOWN;
     }
+    return answers;
+}
+
+static void handle_resolve(int code, char type, int count, int ttl,
+                    void *addresses, void *opaque) {
+
+    auto context = static_cast<QueryContext *>(opaque);
+
+    context->message.error_code = code;
+
+    switch (code) {
+        case DNS_ERR_NONE:
+        case DNS_ERR_FORMAT:
+        case DNS_ERR_SERVERFAILED:
+        case DNS_ERR_NOTEXIST:
+        case DNS_ERR_NOTIMPL:
+        case DNS_ERR_REFUSED:
+        case DNS_ERR_TRUNCATED:
+        case DNS_ERR_NODATA:
+            context->message.rtt = mk::time_now() - context->ticks;
+            break;
+        default:
+            context->message.rtt = 0.0;
+            break;
+    }
+
+    context->message.answers = build_answers_evdns(code, type, count, ttl,
+            addresses);
+    if (context->message.error_code != DNS_ERR_NONE) {
+        context->callback(mk::dns::dns_error(context->message.error_code), context->message);
+    } else {
+        context->callback(NoError(), context->message);
+    }
+
+    delete context;
+}
+
+void query(QueryClass dns_class, QueryType dns_type, std::string name,
+            std::function<void(Error, Message)> cb,
+            Settings settings,
+            Poller *poller) {
+
+    Message message;
+    Query query;
+
+    // XXX add proper exception handling for evbase creation
+    evdns_base *base = create_evdns_base(settings, poller);
```

Comment:
Here we should catch exceptions (because they are mostly caused by user input) and callback. If we want to be precise, we should probably define different exceptions in `create_evdns_base()` and let the exception caused by a failing `evdns_base_new()` propagate because that's "out of memory" and it's not good to continue in that case.